File: content/develop/use-cases/prefetch-cache/nodejs/sync_worker.js
"use strict";

/**
 * Background sync worker for the prefetch-cache demo.
 *
 * An async task drains the primary's change queue and applies each event
 * to Redis through `PrefetchCache.applyChange`. In a real system, the
 * queue is replaced by a CDC pipeline (Redis Data Integration, Debezium,
 * or an equivalent) that tails the primary's binlog/WAL and writes the
 * same shape of events.
 *
 * The worker exposes `pause()` and `resume()` so maintenance paths
 * (`/reprefetch`, `clear()`) can stop event application without tearing
 * the task down. `pause()` blocks until the worker is parked, so the
 * caller knows no apply is in flight by the time it returns.
 */

class SyncWorker {
  constructor({ primary, cache, pollTimeoutMs = 50 } = {}) {
    if (!primary || !cache) {
      throw new Error("primary and cache are required.");
    }
    this.primary = primary;
    this.cache = cache;
    this.pollTimeoutMs = pollTimeoutMs;

    this._stopped = true;
    this._paused = false;
    this._runPromise = null;

    // Promise that resolves when the worker has confirmed it is parked
    // (idle, with no apply in flight). pause() awaits this.
    this._pausedIdlePromise = null;
    this._pausedIdleResolve = null;

    // Promise that resolves when the worker should leave its parked
    // state — either resume() is called or the worker is stopped.
    this._resumePromise = null;
    this._resumeResolve = null;
  }

  start() {
    if (this._runPromise && !this._stopped) {
      return;
    }
    this._stopped = false;
    this._paused = false;
    this._resetIdleSignal();
    this._resetResumeSignal();
    this._runPromise = this._run();
  }

  /**
   * Signal the worker to exit and await its task.
   *
   * If the join times out the worker is wedged inside applyChange; we
   * leave `_runPromise` populated so a subsequent `start()` does not
   * spawn a second worker on top of the orphan.
   */
  async stop(joinTimeoutMs = 2000) {
    this._stopped = true;
    // Wake any parked waiter so the worker can observe the stop flag.
    if (this._resumeResolve) {
      this._resumeResolve();
    }
    if (!this._runPromise) {
      return;
    }
    const timeout = new Promise((resolve) => setTimeout(() => resolve("timeout"), joinTimeoutMs));
    const result = await Promise.race([this._runPromise.then(() => "ok"), timeout]);
    if (result === "ok") {
      this._runPromise = null;
    }
  }

  /**
   * Stop applying events and wait until the worker is parked.
   *
   * Returns `true` once the worker has confirmed it is idle, or `false`
   * if the timeout elapsed first. While paused, change events accumulate
   * in the primary's queue and are applied in order after `resume()`.
   */
  async pause(timeoutMs = 2000) {
    this._resetIdleSignal();
    this._paused = true;
    if (!this._runPromise || this._stopped) {
      return true;
    }
    const timeout = new Promise((resolve) => setTimeout(() => resolve(false), timeoutMs));
    const idle = this._pausedIdlePromise.then(() => true);
    return Promise.race([idle, timeout]);
  }

  resume() {
    this._paused = false;
    // Wake the parked worker.
    if (this._resumeResolve) {
      this._resumeResolve();
    }
    this._resetResumeSignal();
    this._resetIdleSignal();
  }

  _resetIdleSignal() {
    this._pausedIdlePromise = new Promise((resolve) => {
      this._pausedIdleResolve = resolve;
    });
  }

  _resetResumeSignal() {
    this._resumePromise = new Promise((resolve) => {
      this._resumeResolve = resolve;
    });
  }

  async _run() {
    while (!this._stopped) {
      if (this._paused) {
        // Park until resume() or stop() is called.
        // Resolve `_pausedIdleResolve` on every iteration (after each
        // resume/pause cycle that's re-armed by `_resetIdleSignal`) so a
        // *new* pause() that arrives while we are still parked from the
        // previous cycle gets acknowledged immediately, not after the
        // caller's full pause-timeout.
        while (this._paused && !this._stopped) {
          if (this._pausedIdleResolve) {
            this._pausedIdleResolve();
          }
          await this._resumePromise;
        }
        // Loop will re-check flags at the top.
        continue;
      }

      const change = await this.primary.nextChange(this.pollTimeoutMs);
      if (change === null || change === undefined) {
        continue;
      }
      try {
        await this.cache.applyChange(change);
      } catch (err) {
        // Demo behaviour: log and drop the event. A production CDC
        // consumer would retry with bounded backoff and expose a
        // dead-letter / error counter; see the guide's "Production
        // usage" section.
        console.error(`[sync] failed to apply ${JSON.stringify(change)}: ${err && err.message}`);
      }
    }
  }
}

module.exports = { SyncWorker };

File: content/develop/use-cases/prefetch-cache/ruby/sync_worker.rb
# Background sync worker for the prefetch-cache demo.
#
# A daemon thread drains the primary's change queue and applies each
# event to Redis through +PrefetchCache#apply_change+. In a real system
# the queue is replaced by a CDC pipeline (Redis Data Integration,
# Debezium, or an equivalent) that tails the primary's binlog/WAL and
# writes the same shape of events.
#
# The worker exposes +pause+ and +resume+ so maintenance paths
# (+/reprefetch+, +clear+) can stop event application without tearing
# the thread down. +pause+ blocks until the worker is parked, so the
# caller knows no apply is in flight by the time it returns.

require "thread"

class SyncWorker
  # Drain primary change events into Redis on a daemon thread.

  def initialize(primary:, cache:, poll_timeout_s: 0.05)
    @primary = primary
    @cache = cache
    @poll_timeout_s = poll_timeout_s

    @state_mutex = Mutex.new
    @state_cv = ConditionVariable.new
    @stop = false
    @pause = false
    @paused_idle = false
    @thread = nil
  end

  def start
    @state_mutex.synchronize do
      return if @thread && @thread.alive?
      @stop = false
      @pause = false
      @paused_idle = false
    end
    @thread = Thread.new { run_loop }
    @thread.name = "prefetch-cache-sync" if @thread.respond_to?(:name=)
    @thread
  end

  # Signal the worker to exit and join its thread.
  #
  # If the join times out the worker is wedged inside +apply_change+;
  # we leave +@thread+ populated so a subsequent +start+ does not spawn
  # a second worker on top of the orphan.
  def stop(join_timeout_s: 2.0)
    @state_mutex.synchronize do
      @stop = true
      @state_cv.broadcast
    end
    thread = @thread
    return if thread.nil?
    thread.join(join_timeout_s)
    @thread = nil unless thread.alive?
  end

  # Stop applying events and block until the worker is parked.
  #
  # Returns +true+ once the worker has confirmed it is idle, or
  # +false+ if the timeout elapsed first. While paused, change events
  # accumulate in the primary's queue and are applied in order after
  # +resume+.
  def pause(timeout_s: 2.0)
    deadline = monotonic + timeout_s
    @state_mutex.synchronize do
      @pause = true
      @paused_idle = false
      @state_cv.broadcast
      return true if @thread.nil? || !@thread.alive?
      until @paused_idle
        remaining = deadline - monotonic
        return false if remaining <= 0
        @state_cv.wait(@state_mutex, remaining)
      end
      true
    end
  end

  def resume
    @state_mutex.synchronize do
      @pause = false
      @paused_idle = false
      @state_cv.broadcast
    end
  end

  private

  def run_loop
    loop do
      should_stop, should_pause = @state_mutex.synchronize { [@stop, @pause] }
      break if should_stop

      if should_pause
        @state_mutex.synchronize do
          # Park until the pause is lifted or the worker is stopped.
          # Re-set @paused_idle on every iteration so a *new* pause
          # that arrives while we are still parked from the previous
          # cycle gets acknowledged immediately, not after the caller's
          # full pause-timeout.
          while @pause && !@stop
            @paused_idle = true
            @state_cv.broadcast
            @state_cv.wait(@state_mutex, @poll_timeout_s)
          end
          @paused_idle = false
        end
        next
      end

      change = @primary.next_change(@poll_timeout_s)
      next if change.nil?
      begin
        @cache.apply_change(change)
      rescue => exc
        # Demo behaviour: log and drop the event. A production CDC
        # consumer would retry with bounded backoff and expose a
        # dead-letter / error counter; see the guide's "Production
        # usage" section.
        warn "[sync] failed to apply #{change.inspect}: #{exc}"
      end
    end
  end

  def monotonic
    Process.clock_gettime(Process::CLOCK_MONOTONIC)
  end
end
